### PR TITLE
fix ActorPublisher state machine, fixes #20031

### LINKED ISSUE #20031: FileIO publisher invokes onComplete twice
Akka version: 2.4.2
OS: Ubuntu 15.10

`FileIO` violates section 1.7 of the reactive streams spec, *Once a terminal state has been signaled (onError, onComplete) it is REQUIRED that no further signals occur.*

Reproducer:

```scala
import akka.actor._
import akka.util.ByteString
import akka.stream._
import akka.stream.scaladsl._
import org.reactivestreams._

implicit val system = ActorSystem()
implicit val mat = ActorMaterializer()
val file = new java.io.File(/* insert any existing file here */ "/home/jroper/bike.jpg")
val subscriber = new Subscriber[ByteString] {
  var s: Subscription = _
  def onSubscribe(s: Subscription) = {
    println("onSubscribe")
    this.s = s
    s.request(1)
  }
  def onNext(b: ByteString) = s.request(1)
  def onError(t: Throwable) = println(t)
  def onComplete = println("onComplete")
}

FileIO.fromFile(file).runWith(Sink.fromSubscriber(subscriber))
```

Running it:

```
scala> FileIO.fromFile(file).runWith(Sink.fromSubscriber(subscriber))
res0: akka.NotUsed = NotUsed

scala> onSubscribe
onComplete
onComplete
```

### PR DESCRIPTION
**Pull request by rkuhn**
ref #20031

### CODE REVIEW

File: akka-stream/src/main/scala/akka/stream/impl/io/InputStreamPublisher.scala [hunk 1]
```diff
@@ -47,7 +47,7 @@ private[akka] class InputStreamPublisher(is: InputStream, completionPromise: Pro
   def readAndSignal(): Unit =
     if (isActive) {
       readAndEmit()
-      if (totalDemand > 0) self ! Continue
+      if (totalDemand > 0 && isActive) self ! Continue
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/impl/io/InputStreamPublisher.scala]: Ah, not forgetting that one is tricky as hell

Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/impl/io/InputStreamPublisher.scala]: to be fair: this only gets rid of a dead letter, but people don’t really love those somehow …

**Comment by ktoso** [comment]
LGTM AFAICS

File: akka-stream/src/main/scala/akka/stream/impl/io/InputStreamPublisher.scala (hunk 1 above)
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/io/InputStreamPublisher.scala]: I think it's a name-thing

**Comment by patriknw** [comment]
LGTM

**Pull request merged**